DOCSP-42319: Read and write settings (#79)
mongodb/docs-scala: source/read-write-pref.txt

.. _scala-read-write-pref:

====================================
Configure Operations on Replica Sets
====================================

.. facet::
   :name: genre
   :values: reference

.. meta:: 
   :keywords: customize, preferences, replica set, consistency

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to configure **write concern**, **read concern**,
and **read preference** options to modify the way that the {+driver-short+} runs create,
read, update, and delete (CRUD) operations on replica sets.

You can set write concern, read concern, and read preference options at the following
levels:

- Client, which sets the *default for all operation executions* unless overridden
- Transaction
- Database
- Collection

This list also indicates the increasing order of precedence of the option settings. For
example, if you set a read concern level for a transaction, it will override a read
concern level inherited from the client.

> Write concern, read concern, and read preference options allow you to customize the

Author: Yes - by "options" I mean the different read/write concern and read preference options, like primary, majority, secondary, etc

causal consistency and availability of the data in your replica sets. To see a full
list of these options, see the following guides in the {+mdb-server+} manual:

- :manual:`Read Preference </core/read-preference/>`
- :manual:`Read Concern </reference/read-concern/>`
- :manual:`Write Concern </reference/write-concern/>`

> 

Author: The links above go to the read concern, write concern, and read preference pages

on a replica set by setting read and write concerns.

This section shows how to configure these read and write options at the following levels:

- :ref:`Client <scala-read-write-client>`: Specify options in a ``MongoClientSettings`` instance
  or in the connection URI
- :ref:`Transaction <scala-read-write-transaction>`: Specify options in a ``TransactionOptions``
  instance
- :ref:`Database <scala-read-write-database>`: Call corresponding option methods on a
  ``MongoDatabase`` instance
- :ref:`Collection <scala-read-write-collection>`: Call corresponding option methods on a
  ``MongoCollection`` instance

.. _scala-read-write-client:

Client Configuration
~~~~~~~~~~~~~~~~~~~~

This example shows how to set the read preference, read concern, and
write concern of a ``MongoClient`` instance by passing a ``MongoClientSettings``
instance to the constructor. The code configures the following settings:

- ``secondary`` read preference: Read operations retrieve data from 
  secondary replica set members
- ``LOCAL`` read concern: Read operations return the instance's most recent data 
  without guaranteeing that the data has been written to a majority of the replica
  set members

> - ``W2`` write concern: The primary and one secondary replica set member
>   must acknowledge the write operation

Reviewer: S:

```suggestion
- ``W2`` write concern: The primary replica set  member and one secondary member
  must acknowledge the write operation
```

.. literalinclude:: /includes/read-write-pref.scala
    :language: scala
    :dedent:
    :start-after: start-client-settings
    :end-before: end-client-settings

Alternatively, you can specify the read and write settings in the connection
URI, which is passed as a parameter to the ``MongoClient`` constructor:

.. literalinclude:: /includes/read-write-pref.scala
    :language: scala
    :dedent:
    :start-after: start-client-settings-uri
    :end-before: end-client-settings-uri

.. _scala-read-write-transaction:

Transaction Configuration
~~~~~~~~~~~~~~~~~~~~~~~~~

This example shows how to set the read preference, read concern, and
write concern of a transaction by passing a ``TransactionOptions``

> instance to the ``startTransaction()`` method. Transactions run within
> sessions, or groupings of related read or write operations that you
> intend to run sequentially. Before configuring transaction options, create a

Reviewer: S: using "or" to introduce a definition can be confusing

```suggestion
instance to the ``startTransaction()`` method. Transactions run within
*sessions*, groupings of related read or write operations that you
intend to run sequentially. Before configuring transaction options, create a
```

``ClientSession`` instance to start a session.

.. tip::

   To learn more about sessions, see :manual:`Server Sessions </reference/server-sessions/>`
   in the {+mdb-server+} manual.

> The example configures the following settings:

Reviewer: S:

```suggestion
The following code example configures the following settings:
```

Author: I'll leave it as "example" since I don't like the repetition of "following" personally

> - ``primary`` read preference: Read operations retrieve data from 
>   the primary replica set member
> - ``MAJORITY`` read concern: Read operations return the instance's most recent data 
>   that has been written to a majority of replica set members
> - ``W1`` write concern: The primary replica set member must acknowledge the
>   write operation

Reviewer: ```suggestion
- ``primary`` read preference: Read operations retrieve data from 
  the primary replica set member.
- ``MAJORITY`` read concern: Read operations return the instance's most recent data 
  that has been written to a majority of replica set members.
- ``W1`` write concern: The primary replica set member must acknowledge the
  write operation.
```

.. literalinclude:: /includes/read-write-pref.scala
    :language: scala
    :dedent:
    :start-after: start-transaction-settings
    :end-before: end-transaction-settings

.. _scala-read-write-database:

Database Configuration
~~~~~~~~~~~~~~~~~~~~~~

This example shows how to set the read preference, read concern, and

> write concern of a database called ``test_database`` by chaining their setter

Reviewer: S: "their" makes it seems like the methods belong to these options, which is, like, maybe sort of true?

```suggestion
write concern of a database called ``test_database`` by chaining setter
```

...or something similar

methods to the ``getDatabase()`` method. The code configures the following
settings:

> - ``primaryPreferred`` read preference: Read operations retrieve data from 
>   the primary replica set member, or secondary members if the primary is unavailable
> - ``AVAILABLE`` read concern: Read operations return the instance's most recent data 
>   without guaranteeing that the data has been written to a majority of the replica
>   set members
> - ``MAJORITY`` write concern: The majority of all replica set members
>   must acknowledge the write operation

Reviewer: ```suggestion
- ``primaryPreferred`` read preference: Read operations retrieve data from 
  the primary replica set member, or secondary members if the primary is unavailable.
- ``AVAILABLE`` read concern: Read operations return the instance's most recent data 
  without guaranteeing that the data has been written to a majority of the replica
  set members.
- ``MAJORITY`` write concern: The majority of all replica set members
  must acknowledge the write operation.
```

.. literalinclude:: /includes/read-write-pref.scala
    :language: scala
    :dedent:
    :start-after: start-database-settings
    :end-before: end-database-settings

.. _scala-read-write-collection:

Collection Configuration
~~~~~~~~~~~~~~~~~~~~~~~~

This example shows how to set the read preference, read concern, and

> write concern of a collection called ``test_collection`` by chaining their setter

Reviewer: ```suggestion
write concern of a collection called ``test_collection`` by chaining setter
```

methods to the ``getCollection()`` method. The code configures the following
settings:

- ``secondaryPreferred`` read preference: Read operations retrieve data from 

>   secondary replica set members, or the primary members if no secondaries are available

Reviewer: ```suggestion
  secondary replica set members, or the primary members if no secondary members are available
```

> - ``secondaryPreferred`` read preference: Read operations retrieve data from 
>   secondary replica set members, or the primary members if no secondaries are available
> - ``AVAILABLE`` read concern: Read operations return the instance's most recent data 
>   without guaranteeing that the data has been written to a majority of the replica
>   set members
> - ``UNACKNOWLEDGED`` write concern: Does not request acknowledgment of the write operation

Reviewer: ```suggestion
- ``secondaryPreferred`` read preference: Read operations retrieve data from 
  secondary replica set members, or the primary members if no secondaries are available.
- ``AVAILABLE`` read concern: Read operations return the instance's most recent data 
  without guaranteeing that the data has been written to a majority of the replica
  set members.
- ``UNACKNOWLEDGED`` write concern: Does not request acknowledgment of the write operation.
```

.. literalinclude:: /includes/read-write-pref.scala
    :language: scala
    :dedent:
    :start-after: start-collection-settings
    :end-before: end-collection-settings

.. _scala-read-write-advanced:

Advanced Read Configurations 
----------------------------

This section shows how to further customize your read operation
settings in the following ways:

- :ref:`Apply a tag set <scala-tag-sets>`

> - :ref:`Specify a local threshold <scala-local-threshold>`

Reviewer: would probably delete for same reason as above

.. _scala-tag-sets:

Tag Sets
~~~~~~~~

> The {+mdb-server+} allows you to apply key-value :manual:`tags
> </core/read-preference-tags/>` to replica set members
> according to any criteria you choose. You can then use those

Reviewer: S: a little simpler

```suggestion
In {+mdb-server+}, you can apply key-value :manual:`tags
</core/read-preference-tags/>` to replica set members
according to any criteria you choose. You can then use those
```

tags to target one or more members for a read operation.

By default, the {+driver-short+} ignores tags when choosing a member
to read from. To instruct the {+driver-short+} to prefer certain tags,

> pass them as a list to your read preference setter method.

Reviewer: ```suggestion
pass the tags as a list to your read preference setter method.
```

Suppose you are connected to a replica set that contains members hosted
at multiple data centers across the United States. This code example sets
the read preference of ``test_database`` to a tag set that instructs the database
to prefer reads from secondary replica set members in the following order:

1. Members from the New York data center (``Tag("dc", "ny")``)
#. Members from the San Francisco data center (``Tag("dc", "sf")``)

> #. Any secondary members (``TagSet()``)

Reviewer: S: what if you moved this up after the first sentence to make it more linear?

1. suppose you have this replica set
2. the members are tagged like this
3. here's how you'd specify the read order:

Author: Okay I think I reordered similar to what you had in mind

.. literalinclude:: /includes/read-write-pref.scala
   :language: scala
   :dedent:
   :start-after: start-tag-set
   :end-before: end-tag-set

.. _scala-local-threshold:

Local Threshold
~~~~~~~~~~~~~~~

> If multiple replica set members match the read preference and tag sets you specify,

Reviewer: ```suggestion
If multiple replica set members match the read preference and tag sets that you specify,
```

the {+driver-short+} reads from the nearest replica set members, chosen according to
their ping time.

By default, the driver uses only members whose ping times are within 15 milliseconds

> of the nearest member for queries. To distribute reads between members with

Reviewer: I: [between](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#a:~:text=between-,When%20referring%20to%20three%20or%20more%20persons%20or%20items%20or%20when%20the%20number%20is%20unspecified%2C%20use%20among.,-When%20referring%20to)

```suggestion
of the nearest member for queries. To distribute reads among members with
```

higher latencies, specify the ``localThresholdMS`` option in your connection URI.

The following example specifies a local threshold of 35 milliseconds:

.. literalinclude:: /includes/read-write-pref.scala
    :language: scala
    :dedent:
    :start-after: start-local-threshold
    :end-before: end-local-threshold

In the preceding example, the {+driver-short+} distributes reads among matching members
within 35 milliseconds of the closest member's ping time.

.. note::
  

>    The {+driver-short+} ignores the value of ``localThresholdMS`` when communicating with a

Reviewer: S:

```suggestion
   The {+driver-short+} ignores the ``localThresholdMS`` option when communicating with a
```
